### ISSUE TITLE
How to get transaction log, please give some help, thank you!

### ISSUE THREAD

**Issue by dengbangquan** [Boilertalk/Web3.swift#154]
How do I get all the transaction logs in my wallet? I don't know if that's the right way to write it.

![image](https://user-images.githubusercontent.com/16422981/222041041-1141c5e7-3550-47cc-b474-8e41ca30f61a.png)

**Comment by Florian-S-A-W** [comment]
You can use the getLogs function like this:

```
let logs = try await web3.eth.getLogs(
                    addresses: nil,
                    topics: topics,
                    fromBlock: .block(block),
                    toBlock: .block(block)
                )
```

This would get all event logs for the topics given on the specified block.
You could define the topics like this to fetch logs for all Mint Events on any Uniswap V3 pool: 

```
let topics: [[EthereumData]] = [
   [
        // UniswapV3 Mint
        try! EthereumData(
            ethereumValue: "0x7a53080ba414158be7ec69b987b5fb7d07dee101fe85488f0853ae16239d0bde"
        )
]
]
```
What exactly are you trying to do?

**Comment by dengbangquan** [comment]
Can this getLogs get all transaction logs? I tried to get only contract transactions but not main chain transactions. How should the topics field be written, thank you!

**Comment by dengbangquan** [comment]
Can this getLogs get all transaction logs? I tried to get only contract transactions but not main chain transactions. How should the topics field be written, thank you!

**Comment by dengbangquan** [comment]
@Florian-S-A-W hello，
![image](https://user-images.githubusercontent.com/16422981/223098186-33ddbc78-1868-457d-929e-673f13008a83.png)
I don't know whether it is correct to define the "topics" parameter in this way, or can you give me some specific implementation? Thank you!

**Comment by Florian-S-A-W** [comment]
Yes this is correct. 
The way the topics are defined in the JSON RPC Standard is a bit confusing as its not well documented in the API documentation in my opinion.
https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_getlogs
The topics are an Array of Arrays because every transaction can have up to 4 topics, the first is always the signature of the Event, in your case Transfer. Each inner array defines one Collection of 4 topics you are filtering for, in order. The way you defined it, you are filtering for all Transactions that have the Transfer Event as a first topic and not discriminating what the other 3 topics are. 
If you want to filter for multiple Events, you need to add them in the outer array.

So basically the inner array acts as an AND filter and the outer array acts as an OR filter.

**Comment by dengbangquan** [comment]
@Florian-S-A-W  Thank you! 
Through eth_getLogs, I can only get the transaction log of the contract, but cannot get the simple log of the main chain. What should I do?

**Comment by koraykoska** [comment]
@dengbangquan What do you mean? Logs are always emitted during transactions. So eth_getLogs searches for logs emitted during transactions in the range of blocks you define. Each log entry has a transaction hash property that shows which transaction the log belongs to.
There are no logs emitted outside of txs.

**Comment by dengbangquan** [comment]
@Florian-S-A-W @koraykoska 
Do I have to use the socket if I want to use subscribeToLogs?

**Comment by dengbangquan** [comment]
@Florian-S-A-W @koraykoska  
![image](https://user-images.githubusercontent.com/16422981/227833912-2d7ee2f4-c689-496e-84f5-9c62ad702cbb.png)
error:providerDoesNotSupportSubscriptions

**Comment by koraykoska** [comment]
You have to use Websocket provider. It’s part of Web3.swift so it should be
a drag & drop replacement. Just use the initializer of Web3 with wsURL
instead of rpcURL and pass a Websocket url. If your provider doesn’t
support Websocket you can use Chainnodes:

https://www.chainnodes.org
